### PR TITLE
Add ability to escape identifiers to prevent accidental nesting

### LINKED ISSUE #2: Escape nesting delimiters in identifiers [@redux-dynostore/core]
### Is it a bug, feature request or question?

Feature request

### Which package(s) does this involve?

@redux-dynostore/core

### Input Code & Current Behaviour

This issue is duplicating ioof-holdings/redux-dynamic-reducer#10 from the old library.

@anndroow says

> Hi,
> 
> In the documentation, there are three examples of using store.attachReducers():
> 
> ```
> store.attachReducers({
>     some: {
>         path: {
>             to: {
>                 dynamicReducer
>             }
>         }
>     }
> } )
> ```
> 
> `store.attachReducers({ 'some.path.to': { dynamicReducer } } } })`
> 
> `store.attachReducers({ 'some/path/to': { dynamicReducer } } } })`
> 
> The last two examples use a string to specify the path and in the last one, slash is used as the delimiter. But when using the first approach to describe the path of a nested location, it seems like slashes in the property names of the path object are treated as path delimiters as well. The thing is that we're using slashes in some of the property names in the state tree.
> 
> Here's an example of what's happening for us:
> 
> ```
> store.attachReducers({
>     routes: {
>         '/route1/section1': dynamicReducer
>     }
> } )
> ```
> 
> ...will end up like this:
> 
> ```
> {
>     routes: {
>         '': {
>             route1/section1': {...}
>         }
>     }
> }
> ```

### Possible Solution

It was proposed by @jpeyper that providing an escape function for the identifiers could resolve the issue:

```js
store.attachReducers({
    routes: {
        [escapeIdentifier('/route1/section1')]: dynamicReducer
    }
} )
```

### PR DESCRIPTION
**Pull request by mpeyper**
| Q                       | A
| ----------------------- | ---
| Fixed Issues?           | Fixes #2
| Documentation only PR   | 
| Patch: Bug Fix?         | 
| Minor: New Feature?     | 👍
| Major: Breaking Change? | 
| Tests Added + Pass?     | Yes

Added 3 new utility functions to the core package:
1. `escapeIdentifier` - escapes `.` and `/` chars in the given string so that they don't trigger nesting logic
2. `escapeIdentifiers` - calls `escapeIdentifiers` on all the keys, and nested keys recursively, of the given object
~3. `unescapeIdentifier` - undoes the escaping of `escapeIdentifiers`

Questions:
~1. Should `unescapeIdentifier` actually be exported?  It's used internally, but not sure if anyone else would actually use it~
2. Should there also be an `unescapeIdentifiers` function? It wasn't needed to implement the escaping logic, but it wouldn't be difficult to add
~3. Should `dynamic` automatically escape the given identifiers?  I have a feeling it's not possible to actually use a component that has a `.` or `/` in the identifier unless it is escaped~

### CODE REVIEW

**Comment by mpeyper** [comment]
> 3. Should dynamic automatically escape the given identifiers? I have a feeling it's not possible to actually use a component that has a . or / in the identifier unless it is escaped

I just realised #3 was raised because it definitely is an issue.  @jpeyper (and anyone else) would you mind commenting on that issue with some thoughts on it.

**Comment by mpeyper** [comment]
> 1. Should `unescapeIdentifier` actually be exported? It's used internally, but not sure if anyone else would actually use it.

I've just been looking at #3 and realised that `@redux-dynostore/react-redux-subspace` will need `unescapeIdentifier` so that the `subspaced` component can find it's state based on the escaped identifier the enhancer is provided.

That means that 2. is the only unresolved question.

File: packages/redux-dynostore-core/src/escapeIdentifier.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import isPlainObject from 'lodash.isplainobject'
+
+const ESCAPE_TABLE = {
+  '.': '%%DOT_TOKEN%%',
```
Comment (jpeyper) [inline packages/redux-dynostore-core/src/escapeIdentifier.js]: Part of me thinks this is overly complex and reduces readability in debugging (question, do you see this in the redux dev tools?)

can we just escape with a preceding `\`, then when splitting on the name, use a regex like `(?<!\\)\.` ?

**Comment by mpeyper** [comment]
Closing this as it's wildely out of date and I think the issue was resolved in v2 anyway (maybe?)

**Pull request closed without merging**